File: core/src/main/java/dev/bee/kanjianki/core/RecentMistakePolicy.java
package dev.bee.kanjianki.core;

public final class RecentMistakePolicy {
    private RecentMistakePolicy() {
    }

    public static int boundedLimit(int limit) {
        return Math.max(1, limit);
    }

    public static String[] mistakeRatings() {
        return new String[]{StudyRatings.AGAIN, StudyRatings.HARD};
    }

    public static RecentMistake mistake(String kanji, String rating, long reviewedAtMillis) {
        return new RecentMistake(kanji, rating, reviewedAtMillis);
    }

    public record RecentMistake(String kanji, String rating, long reviewedAtMillis) {
        public RecentMistake {
            kanji = kanji == null ? "" : kanji;
            rating = rating == null ? "" : rating;
        }
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/RecentMistakePolicy.kt
package dev.bee.kanjianki.core

object RecentMistakePolicy {
    @JvmStatic
    fun boundedLimit(limit: Int): Int {
        return maxOf(1, limit)
    }

    @JvmStatic
    fun mistakeRatings(): Array<String> {
        return arrayOf(StudyRatings.AGAIN, StudyRatings.HARD)
    }

    @JvmStatic
    fun mistake(kanji: String?, rating: String?, reviewedAtMillis: Long): RecentMistake {
        return RecentMistake(kanji.orEmpty(), rating.orEmpty(), reviewedAtMillis)
    }

    class RecentMistake(kanji: String?, rating: String?, reviewedAtMillis: Long) {
        private val kanji = kanji.orEmpty()
        private val rating = rating.orEmpty()
        private val reviewedAtMillis = reviewedAtMillis

        fun kanji(): String = kanji

        fun rating(): String = rating

        fun reviewedAtMillis(): Long = reviewedAtMillis

        override fun equals(other: Any?): Boolean {
            return other is RecentMistake &&
                kanji == other.kanji &&
                rating == other.rating &&
                reviewedAtMillis == other.reviewedAtMillis
        }

        override fun hashCode(): Int {
            var result = kanji.hashCode()
            result = 31 * result + rating.hashCode()
            result = 31 * result + reviewedAtMillis.hashCode()
            return result
        }

        override fun toString(): String {
            return "RecentMistake[kanji=$kanji, rating=$rating, reviewedAtMillis=$reviewedAtMillis]"
        }
    }
}
